src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -16,6 +16,66 @@ namespace Microsoft.CodeAnalysis.CSharp
 {
     partial class Binder
     {
+        private NamedTypeSymbol? PrepareForUnionMatching(SyntaxNode node, ref TypeSymbol inputType, BindingDiagnosticBag diagnostics)
+        {
+            CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+            if (inputType is NamedTypeSymbol named && named.IsUnionTypeWithUseSiteDiagnostics(ref useSiteInfo))
+            {
+                // PROTOTYPE: Check feature availability
+
+                inputType = GetUnionTypeValueProperty(named, ref useSiteInfo)?.Type ?? Compilation.GetSpecialType(SpecialType.System_Object);
+                diagnostics.Add(node, useSiteInfo);
+                return named;
+            }
+
+            diagnostics.Add(node, useSiteInfo);
+            return null;
+        }
+
+        internal static PropertySymbol? GetUnionTypeValueProperty(NamedTypeSymbol inputUnionType, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            Debug.Assert(inputUnionType.IsUnionTypeNoUseSiteDiagnostics);
+
+            var iUnion = inputUnionType.AllInterfacesNoUseSiteDiagnostics.First(static (i) => i.IsWellKnownTypeIUnion());
+
+            PropertySymbol? valueProperty = null;
+            foreach (var m in iUnion.GetMembers(WellKnownMemberNames.ValuePropertyName))
+            {
+                if (m is PropertySymbol prop && HasIUnionValueSignature(prop))
+                {
+                    valueProperty = prop;
+                    break;
+                }
+            }
+
+            if (valueProperty is not null)
+            {
+                Debug.Assert(valueProperty.Type.IsObjectType());
+                useSiteInfo.Add(valueProperty.GetUseSiteInfo());
+            }
+            else
+            {
+                useSiteInfo.AddDiagnosticInfo(new CSDiagnosticInfo(ErrorCode.ERR_MissingPredefinedMember, iUnion, WellKnownMemberNames.ValuePropertyName)); // PROTOTYPE: Cover this code path
+            }
+
+            return valueProperty;
+        }
+
+        public static bool HasIUnionValueSignature(PropertySymbol property)
+        {
+            // PROTOTYPE: Cover individual conditions with tests
+            return property is
+            {
+                IsStatic: false,
+                DeclaredAccessibility: Accessibility.Public,
+                IsAbstract: true,
+                GetMethod: not null,
+                SetMethod: null,
```

Comment:
What part of the spec forbids implicitly implementing `IUnion.Value` with a type that has a `set` or `init` property?

Reply:
> What part of the spec forbids implicitly implementing `IUnion.Value` with a type that has a `set` or `init` property?

The spec doesn't say anything about that, and this function doesn't say anything about that either. It is used only for a property on an `IUnion` interface, which is specified as having the following shape:
```
public interface IUnion
{
    // The value of the union or null
    object? Value { get; }
}
```